### PR TITLE
codegen - support multiple annotation keys

### PR DESCRIPTION
**Pull request by dprotaso**
- allow multiple values for comment tag keys
- add OR filter chain
- update codegen to support multiple annotation keys

Part of: https://github.com/knative/serving/issues/12103

### CODE REVIEW

**Comment by dprotaso** [comment]
/assign @n3wscott @nak3 
/hold cancel

Looks like networking => net-* is good and things build properly

**Comment by nak3** [comment]
/cc @julz @markusthoemmes 

It looks good to me. But I don't have much experience around this code.

File: codegen/cmd/injection-gen/generators/comment_parser.go
Comment (n3wscott) [inline codegen/cmd/injection-gen/generators/comment_parser.go:43]: Can you add a prefix example here too?

File: codegen/cmd/injection-gen/generators/comment_parser.go
Comment (n3wscott) [inline codegen/cmd/injection-gen/generators/comment_parser.go:61]: the comment above seems to think that "marker" + "prefix" is the annotation key now. But I do not see the marker first part getting used?

Oh I see, the marker is not needed, just to find the prefix and tags.

aaah, ok the marker becomes just "+", and prefix is the tag value. 


I think this might give us too many tags? It will return all tags in the file and the original version of this will filter the tags for just what we want. But I am not thinking of a downside to returning all the tags. It just causes the downstream to filter more CommentTags.

**Review by n3wscott — APPROVED** [review]
/lgtm

/hold

I think the comment could be updated to be more clear the `prefix` is "+", but LGTM and unhold if you disagree.

File: codegen/cmd/injection-gen/generators/comment_parser.go
Comment (dprotaso) [reply codegen/cmd/injection-gen/generators/comment_parser.go:43]: The comment is a bit confusing

the marker is `+` the prefix is `foo` and `bar` in the example

Maybe I'll update the comment to make it more obvious

File: codegen/cmd/injection-gen/generators/comment_parser.go
Comment (dprotaso) [reply codegen/cmd/injection-gen/generators/comment_parser.go:61]: > It will return all tags in the file and the original version of this will filter the tags for just what we want.

Not quiet - prior this function would convert

`+foo:key=value1,key=value2` to `foo: {key:value2}` 

now it converts that to `foo: {key:[value1, value2]}`

Comment (dprotaso) [reply codegen/cmd/injection-gen/generators/comment_parser.go:61]: > It just causes the downstream to filter more CommentTags.

To clarify downstream has to iterate over the `CommentTag[key]` array while before it returned a string

File: codegen/cmd/injection-gen/generators/comment_parser.go
Comment (dprotaso) [reply codegen/cmd/injection-gen/generators/comment_parser.go:46]: done

**Comment by n3wscott** [comment]
/lgtm

Thanks Dave!!

**Pull request merged**
